2019/en/src/0xa3-excessive-data-exposure.md

```diff
@@ -42,6 +42,8 @@ the site.
 * Never rely on the client side to perform sensitive data filtering.
 * Review the responses from the API to make sure they contain only legitimate
   data.
+* Explicitly define and enforce data returned by all API methods including errors: give all JSON objects schemas, all string objects patterns, use clear field names.
+* Define all sensitive and personally identifiable information (PII) that your application stores and works with and review all API calls returning such information to see if these responses can be a security issue.
```

Comment:
@PauloASilva, let me try to explain the thinking behind this. In my experience, a lot of A3 happen inadvertently.

For example, the recent Uber vulnerability leaked user UUID in the error message and then this UUID could be used to get full mobile app token in another API call (for getConsentScreenDetails).

The UUID leakage they wouldn't even be able to see in the API definition because the error object defined the message as generic string and UUID happened to be a part of the string. This should not be happening. Instead, when developers ever need to expose any data, they need to return structured object so the fields can be looked at.

The getConsentScreenDetails returned a huge JSON with tons of data - a likely (we do not know for sure, they have not publicly shared the API contract) sign that presumably the API contract didn't have the schema defined. If it had - their sec team would have been able to question each field in there. Point is that *very often* responses and errors are not reviewed because they are not defined beyond just `string` or just `JSON`.

Makes sense?

Reply:
Your original contribution was very clear. No need to make them single bullet.

Let's keep up with your original contribution, with a little change to avoid mentioning JSON but instead "structured data".